sstables/trie/trie_writer.hh

```diff
@@ -0,0 +1,1192 @@
+    size_t allocated_memory() const {
+        return (_old.size() + bool(_current)) * _segment_size;
+    }
+
+    // Allocate a suitably-aligned (uninitialized) buffer for an array of N elements of type T.
+    // The total size of this allocation must be smaller than segment_size.
+    template <typename T>
+    [[nodiscard]] ptr<T> alloc()
+    requires (!std::is_array_v<T>) {
+        return alloc_impl<T>(1);
+    }
+
+    template <typename T>
+    [[nodiscard]] ptr<T> alloc(size_t n)
+    requires (std::is_array_v<T>) {
+        return alloc_impl<T>(n);
+    }
+
+    // Invalidate all bytes in the stack since the byte at given address (inclusive).
+    template <typename T>
+    void discard(ptr<T> p) noexcept {
+        auto addr = reinterpret_cast<std::byte*>(p.get());
+        expensive_log("bump_allocator: discard: this={} addr={} _global_pos={} _remaining={}",
+            fmt::ptr(this), fmt::ptr(addr), global_pos(), _remaining);
+        discard_impl(addr);
+        #if TRIE_SANITIZE_BUMP_ALLOCATOR
+            _discard_timestamp++;
+            _max_legal_timestamp.erase(_max_legal_timestamp.upper_bound(p._global_pos), _max_legal_timestamp.end());
+            _max_legal_timestamp.insert({p._global_pos, _discard_timestamp});
+        #endif
+    }
+};
+
+template <typename T>
+using ptr = bump_allocator::ptr<T>;
+
+// The user of the writer inserts (key, payload) pairs. This is the payload.
+// A payload consists of a 4-bit header and several bytes of content.
+struct trie_payload {
+    // What the reader will see is a 4-bit "payload bits" header, and a pointer to the body ("bytes") of the payload.
+    // The reader will use the header to determine the length and the meaning of the "bytes".
+    //
+    // The meaning of "bits" and "bytes" is generally opaque to the trie-writing layer,
+    // except that the value `bits == 0` must represent "no payload", and "bytes" is assumed
+    // empty in this case.
+    uint8_t _payload_bits = 0;
+    // We store the payload in a std::array + size pair,
+    // since it makes managing allocations slightly easier.
+    //
+    // 20 bytes is the biggest maximum possible payload in the BTI index format,
+    // so we don't need more.
+    std::array<std::byte, 20> _payload_buf = {};
+    // For the writer, the meaning of the "bits" is opaque,
+    // we just store an explicit size of the "bytes".
+    uint8_t _payload_size = 0;
+
+    trie_payload() noexcept = default;
+    trie_payload(uint8_t bits, const_bytes blob) noexcept {
+        expensive_assert(blob.size() <= _payload_buf.size());
+        expensive_assert(bits < 16);
+        expensive_assert(bool(blob.size()) == bool(bits));
+        _payload_bits = bits;
+        _payload_size = blob.size();
+        std::ranges::copy(blob, _payload_buf.data());
+    }
+    bool operator==(const trie_payload& other) const {
+        return other._payload_bits == _payload_bits
+            && std::ranges::equal(other.blob(), blob());
```

Comment:
Likely faster to do a size-independent compare (if trailer is consistently initialized).

Reply:
The only reason to compare payloads is testing. This won't have any use outside of tests.